From 0570d3fbd0ba165a464245f2d85e384837bfd76c Mon Sep 17 00:00:00 2001
From: Nava kishore Manne <nava.manne@xilinx.com>
Date: Thu, 30 Jul 2020 07:00:40 +0530
Subject: [PATCH 1482/1852] fpga: versal: Use the scatterlist interface

commit 5350379cd7127e081f0662fd87b53a566835e998 from
https://github.com/Xilinx/linux-xlnx.git

Allows drivers to request the Configuration image
be loaded from dma-able continuous buffer to avoid
needless memory pressure and delays due to multiple
copies.

Signed-off-by: Nava kishore Manne <nava.manne@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/fpga/versal-fpga.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/drivers/fpga/versal-fpga.c b/drivers/fpga/versal-fpga.c
index a6694d7fd27a..2848dab3117e 100644
--- a/drivers/fpga/versal-fpga.c
+++ b/drivers/fpga/versal-fpga.c
@@ -37,6 +37,22 @@ static int versal_fpga_ops_write_init(struct fpga_manager *mgr,
 	return 0;
 }
 
+static int versal_fpga_ops_write_sg(struct fpga_manager *mgr,
+				    struct sg_table *sgt)
+{
+	const struct zynqmp_eemi_ops *eemi_ops = zynqmp_pm_get_eemi_ops();
+	dma_addr_t dma_addr;
+	int ret;
+
+	if (IS_ERR_OR_NULL(eemi_ops) || !eemi_ops->fpga_load)
+		return -ENXIO;
+
+	dma_addr = sg_dma_address(sgt->sgl);
+	ret = eemi_ops->pdi_load(PDI_SOURCE_TYPE, dma_addr);
+
+	return ret;
+}
+
 static int versal_fpga_ops_write(struct fpga_manager *mgr,
 				 const char *buf, size_t size)
 {
@@ -81,6 +97,7 @@ static const struct fpga_manager_ops versal_fpga_ops = {
 	.state = versal_fpga_ops_state,
 	.write_init = versal_fpga_ops_write_init,
 	.write = versal_fpga_ops_write,
+	.write_sg = versal_fpga_ops_write_sg,
 	.write_complete = versal_fpga_ops_write_complete,
 };
 
-- 
2.31.1

